lighthouse-extension/test/extension-test.js

```diff
@@ -0,0 +1,138 @@
+
+    // start puppeteer
+    browser = await puppeteer.launch({
+      headless: false,
+      executablePath: process.env.CHROME_PATH,
+      args: [
+        `--disable-extensions-except=${lighthouseExtensionPath}`,
+        `--load-extension=${lighthouseExtensionPath}`,
+      ],
+    });
+
+    const page = await browser.newPage();
+    await page.goto('https://www.paulirish.com', {waitUntil: 'networkidle2'});
+    const targets = await browser.targets();
+    const extensionTarget = targets.find(({_targetInfo}) => {
+      return _targetInfo.title === 'Lighthouse' &&
+        _targetInfo.type === 'background_page';
+    });
+
+    if (!extensionTarget) {
+      return await browser.close();
+    }
+
+    const client = await extensionTarget.createCDPSession();
+    const lighthouseResult = await client.send(
+      'Runtime.evaluate',
+      {
+        expression: `runLighthouseInExtension({
+          restoreCleanState: true,
+        }, ${JSON.stringify(lighthouseCategories)})`,
+        awaitPromise: true,
+        returnByValue: true,
+      }
+    );
+
+    if (lighthouseResult.exceptionDetails) {
+      if (lighthouseResult.exceptionDetails.exception) {
+        throw new Error(lighthouseResult.exceptionDetails.exception.description);
+      }
+
+      throw new Error(lighthouseResult.exceptionDetails.text);
+    }
+
+    extensionPage = (await browser.pages())
+      .find(page => page.url().includes('blob:chrome-extension://'));
+  });
+
+  after(async () => {
+    // put the default manifest back
+    fs.writeFileSync(manifestLocation, originalManifest);
+
+    if (browser) {
+      await browser.close();
+    }
+  });
+
+  it('should contain all categories', async () => {
+    const categories = await extensionPage.$$(`#${lighthouseCategories.join(',#')}`);
+    assert.equal(categories.length, lighthouseCategories.length,
+      `${categories.join(' ')} does not match ${lighthouseCategories.join(' ')}`);
+  });
+
+  it('should contain audits of all categories', async () => {
+    for (const category of lighthouseCategories) {
+      let selector = '.lh-audit';
+      let expected = getAuditsOfCategory(category).length;
+      if (category === 'performance') {
+        selector = '.lh-audit,.lh-timeline-metric,.lh-perf-hint';
+        expected = getAuditsOfCategory(category).filter(a => !!a.group).length;
+      }
+
+      const elementCount = await getAuditElementsCount({category, selector});
+
+      assert.equal(expected, elementCount,
+        `${category} does not have the correct amount of audits`);
+    }
+  });
+
+  it('should contain a filmstrip', async () => {
+    const filmstrip = await extensionPage.$('lh-filmstrip');
+
+    assert.equal(null, filmstrip,
+      `filmstrip is not available`);
+  });
+
+  it('should contain no failed audits', async () => {
+    const auditErrors = await extensionPage.$$('.lh-debug');
+    const failedAudits = auditErrors.filter(error =>
```

Comment:
heh. I ran into a similar API misunderstanding. I was doing `$$eval(selector, el => ...`) but its actually `elems` instead of just `el`. Whoopsie daisy.

-----------

try something like this out:


```js
  it('should not have any audit errors', async () => {
    function getDebugStrings(elems) {
      return elems.map(el => ({
        debugString: el.textContent,
        title: el.closest('.lh-audit').querySelector('.lh-score__title').textContent,
      }));
    }

    const auditErrors = await extensionPage.$$eval('.lh-debug', getDebugStrings);
    const errors = auditErrors.filter(item => item.debugString.includes('Audit error:'));
    assert.deepStrictEqual(errors, [], 'Audit errors found within the report');
  });
```

gives a nice purdy error: 
![image](https://user-images.githubusercontent.com/39191/37674935-3345cd36-2c31-11e8-98b9-3575a438e750.png)